### PR TITLE
 *: try to support lock/unlock tables function

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
1. Help to support lock/unlock tables function in TiDB.

2. There are 4 table lock types: `READ`, `READ LOCAL`, `WRITE`, `WRITE LOCAL`. 
Attention: 
* The `WRITE LOCAL` is add in TiDB to support TiFlash usage. Mysql doesn't have `WRITE LOCAL`.
The `WRITE LOCAL` means the session with this lock has write/read permission, and the other session is still has read permission.
* The `READ LOCAL` is not supported in TiDB, only support the syntax currently. Because of we think is it useless.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### CODE REVIEW

**Comment by winkyao** [comment]
Please explain what is `WRITE LOCAL`

File: model/model.go [hunk 1]
```diff
@@ -230,11 +231,101 @@ type TableInfo struct {
...
+	Lock *TableLockInfo `json:"Lock"`
 
 	// Version means the version of the table info.
 	Version uint16 `json:"version"`
 }
 
+// TableLockInfo provides meta data describing a table lock.
+type TableLockInfo struct {
+	Tp TableLockType
+	// Use array because there maybe multiple session both hold same read lock.
```
Comment (bb7133) [inline model/model.go]:
```suggestion
	// Use array because there may be multiple sessions holding the same read lock.
```

File: model/model.go [hunk 2]
```diff
@@ -230,11 +231,101 @@ type TableInfo struct {
...
+	Sessions []SessionInfo
+	State    TableLockState
+	// TS use to record the timestamp of this table lock been public locked.
```
Comment (bb7133) [inline model/model.go]:
```suggestion
	// TS is used to record the timestamp this table lock been locked.
```

File: model/model.go [hunk 3]
```diff
@@ -230,11 +231,101 @@ type TableInfo struct {
...
+}
+
+// SessionInfo contain the session ID and the server ID.
+type SessionInfo struct {
+	ServerID  string
+	SessionID uint64
+}
+
+func (s SessionInfo) String() string {
+	return "server:" + s.ServerID + "_session:" + strconv.FormatUint(s.SessionID, 10)
```
Comment (bb7133) [inline model/model.go]:
```suggestion
	return "server: " + s.ServerID + "_session: " + strconv.FormatUint(s.SessionID, 10)
```

File: model/model.go [hunk 4]
```diff
@@ -230,11 +231,101 @@ type TableInfo struct {
...
+}
+
+// DBTableID is composed by schema ID and table ID.
+type DBTableID struct {
```
Comment (bb7133) [inline model/model.go]: Looks that this is not used anymore, right?

File: model/model.go [hunk 5]
```diff
@@ -230,11 +231,101 @@ type TableInfo struct {
...
+		return "none"
+	}
+}
+
+// TableLockType is the type of the table lock.
+type TableLockType byte
+
+const (
+	TableLockNone TableLockType = iota
+	// TableLockRead means the session with this lock has read permission, other session can't read/write until the lock is released.
```
Comment (bb7133) [inline model/model.go]: IMHO we need to add more detailed explanations. For example, if the locks are exclusive, or can be shared.

Although the behavior of `LOCKS` will be explained in our official document, detailed comments are still useful here for the developers.

Comment (bb7133) [reply model/model.go]: I've checked the implementations, this comment is wrong(`other session can't read/write ...`), please fix it.

**Comment by bb7133** [comment]
We should also add `ErrTableLocked`(from https://github.com/pingcap/tidb/pull/10343) related error definitions at parser side.

File: model/model.go (hunk 5 above)
Comment (crazycs520) [reply model/model.go]: Sorry for my mistakes, PTAL again.

File: model/model.go [hunk 6]
```diff
@@ -230,11 +231,98 @@ type TableInfo struct {
...
+	// TableLockRead means the session with this lock can read the table (but not write it).
+	// Multiple sessions can acquire a READ lock for the table at the same time.
+	// Other sessions can read the table without explicitly acquiring a READ lock.
+	TableLockRead
+	// TableLockReadLocal is not supported.
+	TableLockReadLocal
+	// TableLockWrite means only the session with this lock has write/read permission.
+	// Only the session that holds the lock can access the table. No other session can access it until the lock is released.
+	TableLockWrite
+	// TableLockWriteLocal means the session with this lock has write/read permission, and the other session is still has read permission.
```
Comment (bb7133) [inline model/model.go]:
```suggestion
	// TableLockWriteLocal means the session with this lock has write/read permission, and the other session still has read permission.
```

**Comment by bb7133** [comment]
PTAL @winkyao @zimulala

**Pull request merged**